src/main/java/com/avispl/symphony/dal/infrastructure/management/philips/huebridge/PhilipsHueDeviceCommunicator.java

```diff
@@ -0,0 +1,4123 @@
+			typeDevice.put(PhilipsConstant.DEVICE, new HashMap<>());
+			typeDevice.put(PhilipsConstant.ROOM, new HashMap<>());
+			typeDevice.put(PhilipsConstant.ZONE, new HashMap<>());
+			Map<String, String> device = new HashMap<>();
+			typeDevice.put(PhilipsConstant.DEVICE, new HashMap<>());
+			//handle case value is device
+			List<String> deviceOfRoomMapCopy = new LinkedList<>();
+			deviceOfRoomMapCopy.addAll(roomNameAndIdMap.keySet().stream().collect(Collectors.toList()));
+			List<String> deviceOfZoneMapCopy = zoneNameAndIdMap.keySet().stream().collect(Collectors.toList());
+			if (locationItem.getGroup() != null && locationItem.getItems() != null) {
+				int deviceIndex = 0;
+				for (Group groupItem : locationItem.getItems()) {
+					Optional<Map<String, String>> data = deviceNameAndMapDeviceIdOfRoomMap.values().stream().filter(item -> item.containsValue(groupItem.getId()))
+							.findFirst();
+					if (deviceIndex == 0) {
+						initializeDeviceDropdown(device, PhilipsConstant.DEVICE, deviceNameAndDeviceIdZoneMap.values().size());
+					}
+					if (data.isPresent()) {
+						String values = data.toString();
+						Optional<Entry<String, String>> nameDevice = deviceNameAndDeviceIdZoneMap.entrySet().stream().filter(item -> values.contains(item.getValue())).findFirst();
+						if (nameDevice.isPresent()) {
+							device.put(PhilipsConstant.DEVICE + deviceIndex, nameDevice.get().getKey());
+							deviceIndex++;
+						}
+					}
+				}
+				typeDevice.put(PhilipsConstant.DEVICE, device);
+				automationAndTypeMapOfDeviceAndValue.put(groupName, typeDevice);
+				List<String> deviceOfMap = deviceNameAndDeviceIdZoneMap.keySet().stream().collect(Collectors.toList());
+				addRepeatDayForCreateAutomation(property, stats, advancedControllableProperties, deviceOfMap, device, PhilipsConstant.DEVICE);
+
+				String deviceAdd = groupName + PhilipsConstant.HASH + AutomationEnum.DEVICE_ADD.getName();
+				stats.put(deviceAdd, PhilipsConstant.EMPTY_STRING);
+				advancedControllableProperties.add(createButton(deviceAdd, PhilipsConstant.ADD, PhilipsConstant.ADDING, 0));
+
+				String[] typeDropdown = EnumTypeHandler.getEnumNames(TypeOfDeviceEnum.class);
+				AdvancedControllableProperty typeControlProperty = controlDropdown(stats, typeDropdown, property, PhilipsConstant.DEVICE);
+				addOrUpdateAdvanceControlProperties(advancedControllableProperties, typeControlProperty);
+			} else if (locationItem.getGroup() != null && locationItem.getItems() == null) {
+				//handle case value is room
+				if (PhilipsConstant.ROOM.equalsIgnoreCase(locationItem.getGroup().getType())) {
+					String roomId = locationItem.getGroup().getId();
+					Optional<Entry<String, String>> deviceNameInRoom = roomNameAndIdMap.entrySet().stream().filter(item -> item.getValue().equals(roomId)).findFirst();
+					if (zoneIndex == 0) {
+						initializeDeviceDropdown(device, PhilipsConstant.ROOM, roomNameAndIdMap.values().size());
+					}
+					if (deviceNameInRoom.isPresent()) {
```

Comment:
please consider using Optional specific methods (like ifPresentOrElse), if possible.
There are a couple of other optionals in the code - can you also please check them?

Sometimes it is easier to just use isPresent/get, not to overcomplicate things, but in most cases it would make code more compact and coupled around the optional value, saving the context of the logic.

Reply:
I fixed it